src/wallet/bdk_wallet/wallet.rs

```diff
@@ -25,3 +36,191 @@ pub struct Wallet {
+
+        let mut index = KeychainTxOutIndex::<Keychain>::default();
+
+        // get signers_map & mutate index to contain descriptors
+        let signers_map = initialize_signers_and_update_index(&meta_store, &secp, &mut index)
+            .map_err(NewError::Descriptor)?;
+
+        // initiate indexed_graph with the mutated index.
+        let indexed_graph = IndexedTxGraph::new(index);
+
+        // index now contains the descriptors
+        // get the combined changeset
+        let staged = ChangeSet {
+            chain: chain_changeset,
+            indexed_tx_graph: indexed_graph.initial_changeset(),
+            network: Some(network),
+        };
+
+        Ok(Wallet {
+            signers: signers_map,
+            chain,
+            indexed_graph,
+            stage: staged,
+            meta_store,
+            network,
+            secp,
+            rpc,
+            wallet_file_path,
+        })
+    }
+}
+
+pub fn initialize_signers_and_update_index(
+    meta_store: &MetaStore,
+    secp: &Secp256k1<All>,
+    index: &mut KeychainTxOutIndex<Keychain>,
+) -> Result<HashMap<Keychain, Arc<SignersContainer>>, DescriptorError> {
+    let mut signers_map = HashMap::new();
+
+    for keychain in [Keychain::External, Keychain::Internal] {
+        let keychain_xpriv = meta_store
+            .master_key
+            .derive_priv(
+                secp,
+                &DerivationPath::from_str(keychain.path().as_str()).unwrap(),
+            )
+            .unwrap();
+
+        // create descriptor strings with checksum
+        let descriptor_without_checksum = format!("wpkh({}/*)", keychain_xpriv);
+
+        let descriptor_with_checksum = format!(
+            "{}#{}",
+            descriptor_without_checksum,
+            calc_checksum(&descriptor_without_checksum).unwrap()
+        );
+
+        // parse descriptor
+        let (descriptor, keymap) =
+            Descriptor::<DescriptorPublicKey>::parse_descriptor(secp, &descriptor_with_checksum)
+                .map_err(DescriptorError::Miniscript)?;
+
+        // mutate index to include corresponding descriptors of a keychain
+        let _ = index
+            .insert_descriptor(keychain, descriptor.clone())
+            .map_err(|e| match keychain {
+                Keychain::External => {
+                    panic!("this is the first descriptor we're inserting")
+                }
+                Keychain::Internal => {
+                    use bdk_chain::keychain::InsertDescriptorError;
+                    match e {
+                        InsertDescriptorError::DescriptorAlreadyAssigned { .. } => {
+                            DescriptorError::ExternalAndInternalAreTheSame
+                        }
+                        InsertDescriptorError::KeychainAlreadyAssigned { .. } => {
+                            unreachable!("this is the first time we're assigning internal")
+                        }
+                    }
+                }
```

Comment:
When you have to write code like this, that an indication that you need a global Error enum that can hold all the error variants returned from the business logic. Never write a map_err like this. map_err is only used for cases where you can't/don't want to put the error variant inside the enum, thus only for small edge cases.